iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,125 @@
 # limitations under the License.
 
 from lxml import etree
+from enum import Enum
 import re
 
+XHTML_NS = 'http://www.w3.org/1999/xhtml'
+
+class EscapeMode(Enum):
+    DEFAULT = 0
+    STYLE = 1
+
 class XHTMLSerializer:
 
     # From https://www.w3.org/TR/html401/index/elements.html
-    selfClosableElements = (
+    SELF_CLOSABLE = (
         'area', 'base', 'basefont', 'br', 'col', 'frame', 'hr', 'img', 
         'input', 'isindex', 'link', 'meta', 'param'
     )
 
-    def _expandEmptyTags(self, xml):
-        """
-        Expand self-closing tags
-
-        Self-closing tags cause problems for XHTML documents when treated as
-        HTML.  Tags that are required to be empty (e.g. <br>) are left as
-        self-closing.
-        """
-        for e in xml.iter('*'):
-            m = re.match(r'\{http://www\.w3\.org/1999/xhtml\}(.*)', e.tag)
-            if (m is None or m.group(1) not in XHTMLSerializer.selfClosableElements) and e.text is None:
-                e.text = ''
-
-    def serialize(self, xmlDocument, fout):
-        self._expandEmptyTags(xmlDocument)
-        fout.write(etree.tostring(xmlDocument, method="xml", encoding="utf-8", xml_declaration=True))
+    ENTITIES = {
+        '>': '&gt;',
+        '<': '&lt;',
+        '"': '&quot;',
+        '&': '&amp;'
+        }
+
+    def __init__(self, xml_declaration = True, assume_xhtml = True):
+        self.xml_declaration = xml_declaration
+        self.assume_xhtml = assume_xhtml
+
+    def prefix_sort(self, p):
+        return p if p is not None else '0'
+
+    def tag_as_qname(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+
+        prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace), key = self.prefix_sort)))
```

Comment:
What is the intent of this `sorted`?  If it's not possible to determine the prefix that was actually used(?), I would think we always want the most recently bound prefix.

I guess it won't matter much in practice since most documents will predefine all their namespaces on the root anyway.  I wonder if it's worth maintaining a ns-to-prefix map rather than iterating all?

Reply:
The `sorted` is to ensure that the output documents are stable and predictable (for testing purposes) and to ensure that the default prefix is preferred over an explicit prefix.

I don't think we can get access to the prefix used in the source.

Maintaining an NS to prefix map would be reasonable, as we already detect when the prefix bindings have changed, although keeping track of the most recently bound would be a bit awkward, and wouldn't help with the situation where both prefixes are bound at the same time.

Reply:
Got it, thanks.

For elements, it looks like `n.prefix` works.  For attributes, I also don't see a way to get the original prefix from lxml; how unfortunate.  Anyway, duplicate prefixes are hopefully rare.

Reply:
Ah, not sure how I missed `n.prefix`.  I'll update to use that, assuming it's in scope and bound to the right URI.  Which it should be given that we don't support editing of the document.

For attributes, I realise that there's an issue with the current code which we mustn't use an unprefixed attribute name if the attribute has a namespace, because the default namespace binding doesn't apply to attributes.  I think one of the places where duplicate NS bindings do get used is to provide a default binding for elements, and an explicit binding for attributes (and QName-valued elements).

Reply:
Done in 04c08600ec7621436fcf0e5665e283e23976d94c